### PR TITLE
Add test reference topology for TGL nocodec

### PR DESCRIPTION
**Pull request by xiulipan**
Add FW echo reference to nocodec TPLG on SSP0.
Add internal loopback on SSP0. We can compare echo reference anid
loopback capture stream.
Add HW echo reference with DSM comp on SSP2

Add Keyword Detection pipelines to DMIC

merge DSM HWEchoRefernce from https://github.com/thesofproject/sof/pull/3224

Signed-off-by: Pan Xiuli <xiuli.pan@linux.intel.com>

### CODE REVIEW

**Comment by xiulipan** [comment]
@keyonjie @keqiaozhang Please help to test and review.

File: tools/topology/sof-tgl-nocodec.m4 [hunk 1]
```diff
@@ -108,6 +145,34 @@ DAI_ADD(sof/pipe-dai-capture.m4,
 	PIPELINE_SINK_2, 2, s16le,
 	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
 
+# Capture pipeline 9 from demux on PCM 6 using max 2 channels of s32le.
```
Comment (plbossart) [inline tools/topology/sof-tgl-nocodec.m4:148]: pipeline 7?

File: tools/topology/sof-tgl-nocodec.m4 [hunk 2]
```diff
@@ -148,6 +213,7 @@ PCM_DUPLEX_ADD(Port0, 0, PIPELINE_PCM_1, PIPELINE_PCM_2)
 PCM_DUPLEX_ADD(Port1, 1, PIPELINE_PCM_3, PIPELINE_PCM_4)
 PCM_DUPLEX_ADD(Port2, 2, PIPELINE_PCM_5, PIPELINE_PCM_6)
 PCM_CAPTURE_ADD(DMIC, 6, PIPELINE_PCM_13)
+PCM_CAPTURE_ADD(EchoRef, 3, PIPELINE_PCM_7)
```
Comment (plbossart) [inline tools/topology/sof-tgl-nocodec.m4:216]: FwEchoRef

we need to start differentiating between firmware generated echo reference and hardware-generated echo reference. If we don't we'll always have confusion on the topology.

FwEchoRef tells you the data is provided by firmware
HwEchoRef tells you the data is provided by external hardware.

**Comment by xiulipan** [comment]
@plbossart Update the PR with request. Also rename the tplg to `sof-tgl-nocodec-ci`

File: tools/topology/sof/pipe-volume-demux-playback.m4 [hunk 3]
```diff
@@ -23,13 +23,12 @@ include(`mixercontrol.m4')
 include(`bytecontrol.m4')
 
 # demux Bytes control with max value of 255
-C_CONTROLBYTES(DEMUX, PIPELINE_ID,
+C_CONTROLBYTES(concat(`DEMUX', PIPELINE_ID), PIPELINE_ID,
```
Comment (plbossart) [inline tools/topology/sof/pipe-volume-demux-playback.m4:26]: does this impact userspace, e.g. with a change of control names?

File: tools/topology/CMakeLists.txt [hunk 4]
```diff
@@ -117,6 +117,7 @@ set(TPLGS
 	"sof-cml-rt1011-rt5682\;sof-cml-rt1011-rt5682\;-DPLATFORM=cml\;-DPPROC=volume"
 	"sof-cml-rt1011-rt5682\;sof-cml-eq-rt1011-rt5682\;-DPLATFORM=cml\;-DPPROC=eq-iir-eq-fir-volume"
 	"sof-tgl-nocodec\;sof-tgl-nocodec"
+	"sof-tgl-nocodec-ci\;sof-tgl-nocodec-ci"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt:120]: the name is really terrible. something like sof-tgl-nocodec-fw-eref.ri would be clearer, no?

Comment (xiulipan) [reply tools/topology/CMakeLists.txt:120]: @plbossart  Actually this tplg will include both fw echo reference and hw echo reference

File: tools/topology/sof/pipe-volume-demux-playback.m4 (hunk 3 above)
Comment (xiulipan) [reply tools/topology/sof/pipe-volume-demux-playback.m4:26]: Good question. Let me check

Comment (xiulipan) [reply tools/topology/sof/pipe-volume-demux-playback.m4:26]: The name will have some little difference, but no influence.

**Comment by plbossart** [comment]
I maintain that this is an out-of-control endeavor that should be in a test topology repo, not in the topologies distributed in releases. We can't go on like this adding topology files that are for development purposes, or they belong somewhere else.

**Comment by xiulipan** [comment]
@plbossart How about still name it as nocodec? Even if it contains a lot of _advanced_ feature, it still keep the simple nocodec pipelines

File: tools/topology/sof/pipe-volume-demux-playback.m4 [hunk 5]
```diff
@@ -64,7 +63,7 @@ W_PGA(1, PIPELINE_FORMAT, DAI_PERIODS, 2, playback_pga_conf, SCHEDULE_CORE,
 
 # Mux 0 has 2 sink and source periods.
 W_MUXDEMUX(0, 1, PIPELINE_FORMAT, 2, 2, SCHEDULE_CORE,
-	LIST(`         ', "DEMUX"))
+	LIST(`         ', concat(`DEMUX', PIPELINE_ID)))
```
Comment (keyonjie) [inline tools/topology/sof/pipe-volume-demux-playback.m4:66]: @xiulipan, we might need to change the demux_priv definition where using the pipe-volume-demux-playback.m4 also.

```
diff --git a/tools/topology/sof-apl-demux-pcm512x.m4 b/tools/topology/sof-apl-demux-pcm512x.m4
index 642269827..dd185a137 100644
--- a/tools/topology/sof-apl-demux-pcm512x.m4
+++ b/tools/topology/sof-apl-demux-pcm512x.m4
@@ -49,7 +49,7 @@ define(matrix2, `ROUTE_MATRIX(5,
                             `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
 
 dnl name, num_streams, route_matrix list
-MUXDEMUX_CONFIG(demux_priv, 2, LIST(`  ', `matrix1,', `matrix2'))
+MUXDEMUX_CONFIG(demux_priv_1, 2, LIST(`        ', `matrix1,', `matrix2'))
 
 #
 # Define the pipelines
diff --git a/tools/topology/sof-cml-demux-rt5682.m4 b/tools/topology/sof-cml-demux-rt5682.m4
index 3ea8b03d4..0d9f910ba 100644
--- a/tools/topology/sof-cml-demux-rt5682.m4
+++ b/tools/topology/sof-cml-demux-rt5682.m4
@@ -49,7 +49,7 @@ define(matrix2, `ROUTE_MATRIX(5,
                             `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
 
 dnl name, num_streams, route_matrix list
-MUXDEMUX_CONFIG(demux_priv, 2, LIST(`  ', `matrix1,', `matrix2'))
+MUXDEMUX_CONFIG(demux_priv_1, 2, LIST(`        ', `matrix1,', `matrix2'))
 
 #
 # Define the pipelines
diff --git a/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 b/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4
index 4d0927025..1bbb86eff 100644
--- a/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4
+++ b/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4
@@ -51,7 +51,7 @@ define(matrix2, `ROUTE_MATRIX(4,
                             `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
 
 dnl name, num_streams, route_matrix list
-MUXDEMUX_CONFIG(demux_priv, 2, LIST(`  ', `matrix1,', `matrix2'))
+MUXDEMUX_CONFIG(demux_priv_3, 2, LIST(`        ', `matrix1,', `matrix2'))
 ')
 
 #
diff --git a/tools/topology/sof-tgl-max98357a-rt5682.m4 b/tools/topology/sof-tgl-max98357a-rt5682.m4
index 9b7be92c5..2292486bc 100644
--- a/tools/topology/sof-tgl-max98357a-rt5682.m4
+++ b/tools/topology/sof-tgl-max98357a-rt5682.m4
@@ -52,7 +52,7 @@ define(matrix2, `ROUTE_MATRIX(9,
                             `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
 
 dnl name, num_streams, route_matrix list
-MUXDEMUX_CONFIG(demux_priv, 2, LIST(`  ', `matrix1,', `matrix2'))
+MUXDEMUX_CONFIG(demux_priv_1, 2, LIST(`        ', `matrix1,', `matrix2'))
 
 #
 # Define the pipelines
```

**Comment by keyonjie** [comment]
> I maintain that this is an out-of-control endeavor that should be in a test topology repo, not in the topologies distributed in releases. We can't go on like this adding topology files that are for development purposes, or they belong somewhere else.

The intention of adding this is not for development purpose, but for CI regular testing which will run on every submitted PR, it will cover more regressions finding before PRs being merged, on a single limited CI platform. @plbossart it is important for the CI coverage, IMHO.

**Comment by xiulipan** [comment]
@plbossart As this TPLG is only used for platform validation, we can keep the topology not merge but only use the binary from this PR.

**Comment by plbossart** [comment]
@xiulipan @keyonjie All test topologies should me merged and maintained, otherwise there's no way we can work on them as a team.

That said, my point remains that we have to separate out what it used for testing and what should be packaged and used by distros. It's not right if we don't draw a clear line.

We do the same for kconfigs, we have sof-defconfig and sof-dev-defconfig, and recommend distros don't use the latter. Same for topologies.

**Comment by xiulipan** [comment]
@plbossart That is a good point. Can we merge this PR first as this is a blocking issue with our validation plan.

I will try to do a update to the CMakefiles to add a new build target `make dev` and more most our development topology into this build target.

**Comment by xiulipan** [comment]
@plbossart @lgirdwood @keyonjie @ranj063 @singalsu Please help to review this PR to split the topology list into release and development.
I may misplace some tplg file into wrong list, please also help to check.

This will help us to manage the topologies more easily.

File: tools/topology/CMakeLists.txt [hunk 6]
```diff
@@ -139,22 +110,60 @@ set(TPLGS
 	"sof-jsl-da7219\;sof-jsl-da7219\;-DPLATFORM=jsl"
 	"sof-jsl-da7219\;sof-jsl-da7219-mx98360a\;-DPLATFORM=jsl-dedede"
 	"sof-imx8mp-wm8960\;sof-imx8mp-wm8960"
-	"sof-smart-amplifier-nocodec\;sof-smart-amplifier-nocodec"
 	"sof-jsl-rt5682\;sof-jsl-rt5682-rt1015\;-DPLATFORM=jsl-rt1015"
 	"sof-jsl-rt5682\;sof-jsl-rt5682-mx98360a\;-DPLATFORM=jsl-dedede"
 )
 
+# This list contains development topologies. Used for some feature test.
+# To use them, you may need to rename the tplg file.
```
Comment (lgirdwood) [inline tools/topology/CMakeLists.txt:118]: to match the file name used by the driver ?

File: tools/topology/sof-apl-demux-pcm512x.m4 [hunk 7]
```diff
@@ -49,7 +49,7 @@ define(matrix2, `ROUTE_MATRIX(5,
 			     `BITS_TO_BYTE(0, 0, 0 ,0 ,0 ,0 ,0 ,1)')')
 
 dnl name, num_streams, route_matrix list
-MUXDEMUX_CONFIG(demux_priv, 2, LIST(`	', `matrix1,', `matrix2'))
```
Comment (plbossart) [inline tools/topology/sof-apl-demux-pcm512x.m4:52]: typos and grammar in commit message

File: tools/topology/CMakeLists.txt [hunk 8]
```diff
@@ -118,6 +118,7 @@ set(TPLGS
 	"sof-cml-rt1011-rt5682\;sof-cml-rt1011-rt5682\;-DPLATFORM=cml\;-DPPROC=volume"
 	"sof-cml-rt1011-rt5682\;sof-cml-eq-rt1011-rt5682\;-DPLATFORM=cml\;-DPPROC=eq-iir-eq-fir-volume"
 	"sof-tgl-nocodec\;sof-tgl-nocodec"
+	"sof-tgl-nocodec-ci\;sof-tgl-nocodec-ci"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt]: typo in commit message s/andi/and s/ampilfier/amplifier/

File: tools/topology/sof-tgl-nocodec-ci.m4 [hunk 9]
```diff
@@ -24,6 +24,27 @@ include(`sof/tokens.m4')
...
+# to generate dmic setting with kwd when we have dmic
+# define channel
+define(CHANNELS, `2')
+# define kfbm with volume
+define(KFBM_TYPE, `vol-kfbm')
+# define pcm, pipeline and dai id
+define(DMIC_PCM_48k_ID, `99')
+define(DMIC_PIPELINE_48k_ID, `10')
+define(DMIC_DAI_LINK_48k_ID, `6')
+define(DMIC_DAI_LINK_48k_NAME, `NoCodec-6')
```
Comment (plbossart) [inline tools/topology/sof-tgl-nocodec-ci.m4:37]: didn't we already say that this is a horrible name and this is not aligned with the kernel PR?
Really don't get what the directions are.

File: tools/topology/CMakeLists.txt [hunk 10]
```diff
@@ -139,22 +110,60 @@ set(TPLGS
...
+	"sof-cml-rt5682-kwd\;sof-cml-rt5682-kwd\;-DPLATFORM=cml"
+	"sof-cml-rt5682-kwd\;sof-whl-rt5682-kwd\;-DPLATFORM=whl"
+	"sof-cml-rt5682-kwd\;sof-icl-rt5682-kwd\;-DPLATFORM=icl"
+	"sof-tgl-nocodec-ci\;sof-tgl-nocodec-ci"
+	"sof-smart-amplifier-nocodec\;sof-smart-amplifier-nocodec"
+)
+
 add_custom_target(topologies ALL)
+add_custom_target(topology_release)
+add_custom_target(topology_dev)
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt:155]: I don't think it makes sense to have different targets, just move the test files to different directories.

Comment (xiulipan) [reply tools/topology/CMakeLists.txt:155]: Remove the RFC

**Comment by xiulipan** [comment]
@plbossart @lgirdwood As we have discussed about the topology management about production and development. I will try to update this PR with a simple start up to add a `development` folder.

@plbossart I have changed the topology to remove the dependency on Kernel PR. Do we still need some patch to nocodec machine driver to update the dai names?

As this PR only provide a validation topology for our release, I did not want to involve much dependency in this PR. Can we merge this simple files first. And make alignment about what is missed and we can fix them later to do not block any release.

**Comment by aiChaoSONG** [comment]
SOFCI TEST

File: tools/topology/sof-tgl-nocodec-ci.m4 (hunk 9 above)
Comment (keyonjie) [reply tools/topology/sof-tgl-nocodec-ci.m4]: @plbossart I think the kernel PR was abandoned and closed, so the naming is kept intact.

**Review by lgirdwood — APPROVED** [review]
Thanks @xiulipan for the dev folder. @plbossart ok for you now ?

**Comment by xiulipan** [comment]
ping @plbossart

**Comment by xiulipan** [comment]
@lgirdwood I think we can merge this and I will send following clean up PR for the development topologies

**Comment by keyonjie** [comment]
Since no objection for more than 2 weeks, let's proceed and merge it.

**Pull request merged**